include/tvm/relay/feature.h

```diff
@@ -0,0 +1,170 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file tvm/relay/feature.h
+ * \brief Detect features used in Expr/Module.
+ */
+#ifndef TVM_RELAY_FEATURE_H_
+#define TVM_RELAY_FEATURE_H_
+
+#include <tvm/node/container.h>
+#include <tvm/expr.h>
+#include <bitset>
+
+namespace tvm {
+namespace relay {
+
+/*! \brief Different kinds of relay feature a program might use. */
+enum Feature : int {
```

Comment:
My only concern is that whether we need feature types for each kind of expr or only define a few high-level features like `fGraph`, `fPatternMatch`.

Reply:
I think we should at least have one of each kind, as a pass might simply fail due to not pattern matching against it in the visitor.

Reply:
A pass failing due to missing visitor cases seems like the sort of thing that tests should catch. I agree with @vinx13 that it's probably better for these flags to correspond to high-level features

Reply:
The idea is that there might be two different characteristic for a pass: not matching against case x because it is unsupported, and not matching against case x because someone forgot to. We can use the f flag to detect such case, then use it to generate/pull all relay sample program satisfying the feature, and finally run the test automatically, instead of writing basically the same test over and over.